### PR TITLE
azure webhook

### PR DESCRIPTION
**Pull request by mrT23**
## **Type**
enhancement, documentation

___

## **Description**
- Implemented an Azure DevOps Server webhook server with basic authentication and support for "Pull request created" and "Pull request commented on" events.
- Enhanced the Azure DevOps provider with features like comment editing, PR description truncation for long descriptions, and caching for diff files retrieval.
- Added documentation for setting up Azure DevOps webhooks, including security considerations for basic authentication.
- Minor adjustments and enhancements across the codebase, including a debug logging change and Dockerfile update for the new webhook server.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Enhance YAML Loading Functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/algo/utils.py

<li>Adjusted <code>load_yaml</code> function to strip additional trailing colon and <br>newline characters from the response text.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/667/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>azuredevops_provider.py</strong><dd><code>Enhancements and Fixes in Azure DevOps Provider</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/git_providers/azuredevops_provider.py

<li>Introduced <code>MAX_PR_DESCRIPTION_AZURE_LENGTH</code> constant to handle Azure PR <br>description length limit.<br> <li> Added <code>edit_comment</code> method to update comments in Azure DevOps.<br> <li> Modified <code>get_repo_settings</code> to return the first item from the settings <br>list.<br> <li> Implemented caching for <code>get_diff_files</code> to avoid repeated API calls.<br> <li> Added logic to truncate PR descriptions exceeding Azure's length <br>limit.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/667/files#diff-1a90ea92bcfba3cf9f1dbc298d2e570566f4c439cb3650ee746cebdb02ca4a0b">+41/-6</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>azuredevops_server_webhook.py</strong><dd><code>Implement Azure DevOps Server Webhook Server</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/servers/azuredevops_server_webhook.py

<li>Added a new Azure DevOps Server webhook server implementation.<br> <li> Implemented basic authentication for webhook security.<br> <li> Added support for processing "Pull request created" and "Pull request <br>commented on" events.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/667/files#diff-4525cb1c9c5ffe691dee1db5501c614ab50388fdf711220ee18db18cf64a7358">+136/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pr_description.py</strong><dd><code>Minor Debug Logging Adjustment</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/tools/pr_description.py

- Commented out a debug log line for the full markdown description.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/667/files#diff-66130451adce278a098a5770e2db52b12050205fcd4d80c13f813615d4449abb">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Dockerfile</strong><dd><code>Add Dockerfile Stage for Azure DevOps Webhook Server</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
docker/Dockerfile

- Added a new Dockerfile stage for the Azure DevOps webhook server.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/667/files#diff-f34da55ca08f1a30591d8b0b3e885bcc678537b2a9a4aadea4f190806b374ddc">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>Usage.md</strong><dd><code>Document Azure DevOps Webhook Setup and Security</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
Usage.md

<li>Added documentation for setting up and securing Azure DevOps webhooks.<br> <br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/667/files#diff-98bbb2166f184dd69bb689271547bf3414ce37f682b1f51bd0efdb8b565ce99d">+13/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Configuration changes
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>.secrets_template.toml</strong><dd><code>Update Secrets Template with Azure DevOps Credentials</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
pr_agent/settings/.secrets_template.toml

<li>Added template entries for Azure DevOps and Azure DevOps Server <br>webhook credentials.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/667/files#diff-f9cf1bb39fd1a7e126b4d06af0238b24125e183f540c20360a392ef695682783">+11/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

File: pr_agent/servers/azuredevops_server_webhook.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    )
+
+@router.get("/")
+async def root():
+    return {"status": "ok"}
+    
+def start():
+    app = FastAPI(middleware=[Middleware(RawContextMiddleware)])
+    app.include_router(router)
+    uvicorn.run(app, host="0.0.0.0", port=int(os.environ.get("PORT", "3000")))
```
Comment (mrT23) [inline pr_agent/servers/azuredevops_server_webhook.py:132]: /ask explain this logic

File: pr_agent/servers/azuredevops_server_webhook.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+# for this reason, https must be enabled to ensure the credentials are not sent in clear text
+def authorize(credentials: HTTPBasicCredentials = Depends(security)):
+        is_user_ok = secrets.compare_digest(credentials.username, WEBHOOK_USERNAME)
+        is_pass_ok = secrets.compare_digest(credentials.password, WEBHOOK_PASSWORD)
+        if not (is_user_ok and is_pass_ok):
+            raise HTTPException(
+                status_code=status.HTTP_401_UNAUTHORIZED,
+                detail='Incorrect username or password.',
+                headers={'WWW-Authenticate': 'Basic'},
+            )
```
Comment (hussam789) [inline pr_agent/servers/azuredevops_server_webhook.py:55]: /ask make sure to use https be enabled to ensure the credentials are not sent in clear text

File: pr_agent/git_providers/azuredevops_provider.py [hunk 3]
```diff
@@ -124,6 +125,19 @@ def publish_code_suggestions(self, code_suggestions: list) -> bool:
...
+                repository_id=self.repo_slug,
+                pull_request_id=self.pr_num,
+                thread_id=comment["thread_id"],
+                comment_id=comment["comment_id"],
+                comment=Comment(content=body),
+                project=self.workspace_slug,
+            )
+        except Exception as e:
+            get_logger().exception(f"Failed to edit comment, error: {e}")
+
```
Comment (almog-lv) [inline pr_agent/git_providers/azuredevops_provider.py:140]: /ask what this code does?

**Comment by mrT23** [comment]
## PR Code Suggestions ✨

<table><thead><tr><td>Category</td><td align=left>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </td><td align=center>Score</td></tr><tbody><tr><td rowspan=1><strong>Maintainability</strong></td>
<td>



<details><summary>Simplify string manipulation by removing redundant method calls</summary>

___


**The method <code>strip().rstrip()</code> is redundant since <code>strip()</code> already removes whitespace from <br>both the start and the end of the string. Consider simplifying the method chain.**
    
[pr_agent/algo/utils.py [364]](https://github.com/Codium-ai/pr-agent/pull/667/files#diff-6b9df72d53c6f0d89fb142c210238a276c0782305e0024d16fbfcaf72c2e2b53R364-R364)

```diff
-response_text_copy = response_text.strip().rstrip().removeprefix('{').removesuffix('}').rstrip(':\n')
+response_text_copy = response_text.strip().removeprefix('{').removesuffix('}').rstrip(':\n')
 
```
                   


<details><summary><b>Importance[1-10]: 7</b></summary>

Why: The suggestion correctly identifies that the method chain `strip().rstrip()` is redundant and simplifies it, improving code readability and maintainability.


</details></td><td align=center>

7</details></td></tr><tr><td rowspan=2><strong>Possible bug</strong></td>
<td>



<details><summary>Correct the string slicing to avoid exceeding the maximum length limit</summary>

___


**When truncating the PR description, ensure that the final length does not exceed the <br>maximum allowed by Azure DevOps by correctly calculating the remaining length for the <br>truncation message.**
    
[pr_agent/git_providers/azuredevops_provider.py [360]](https://github.com/Codium-ai/pr-agent/pull/667/files#diff-1a90ea92bcfba3cf9f1dbc298d2e570566f4c439cb3650ee746cebdb02ca4a0bR360-R360)

```diff
-pr_body = pr_body[:MAX_PR_DESCRIPTION_AZURE_LENGTH - len(trunction_message)] + trunction_message
+pr_body = pr_body[:MAX_PR_DESCRIPTION_AZURE_LENGTH - len(trunction_message) - 1] + trunction_message
 
```
                   


<details><summary><b>Importance[1-10]: 6</b></summary>

Why: The suggestion correctly identifies a potential off-by-one error in the string slicing for truncating the PR description. This is a valid bug fix, ensuring the length does not exceed the defined maximum.

</details></td><td align=center>

6</details></td></tr><tr><td>



<details><summary>Ensure the method returns cached data correctly by checking for <code>None</code></summary>

___


**The method <code>get_diff_files</code> should return a cached list of diff files if it exists to avoid <br>redundant API calls. This can be achieved by checking <code>self.diff_files</code> at the beginning of <br>the method.**
    
[pr_agent/git_providers/azuredevops_provider.py [224-225]](https://github.com/Codium-ai/pr-agent/pull/667/files#diff-1a90ea92bcfba3cf9f1dbc298d2e570566f4c439cb3650ee746cebdb02ca4a0bR224-R225)

```diff
-if self.diff_files:
+if self.diff_files is not None:
     return self.diff_files
 
```
                   


<details><summary><b>Importance[1-10]: 0</b></summary>

Why: The existing code already correctly checks if `self.diff_files` is truthy before returning it, which implicitly covers the `None` check. The suggestion is unnecessary and does not improve the code.


</details></td><td align=center>

0</details></td></tr><tr><td rowspan=1><strong>Best practice</strong></td>
<td>



<details><summary>Enhance error logging by including the exception message</summary>

___


**When handling exceptions, it's beneficial to log the exception message along with the <br>stack trace for better debugging. Use <code>str(e)</code> to include the exception message in the log.**
    
[pr_agent/git_providers/azuredevops_provider.py [139]](https://github.com/Codium-ai/pr-agent/pull/667/files#diff-1a90ea92bcfba3cf9f1dbc298d2e570566f4c439cb3650ee746cebdb02ca4a0bR139-R139)

```diff
-get_logger().exception(f"Failed to edit comment, error: {e}")
+get_logger().exception(f"Failed to edit comment, error: {str(e)}")
 
```
                   


<details><summary><b>Importance[1-10]: 5</b></summary>

Why: The suggestion to include the exception message in the log is correct and enhances error logging, but it's a minor improvement as the exception message might already be included by default in `exception()` method of logging.


</details></td><td align=center>

5</details></td></tr></tr></tbody></table>

**Comment by mrT23** [comment]
/improve
--pr_code_suggestions.focus_only_on_problems=false

**Pull request merged**